### ISSUE TITLE
After package download error, cache goes bad

### ISSUE THREAD

**Issue by nwolverson** [purescript/spago#933]
A missing package in the registry was causing a download error

```
Couldn't fetch package, status was not ok (StatusCode 404), got answer:
  <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><BucketName>purescript-registry</BucketName><RequestId>tx000000000000019623f28-0063caa37d-34e484cb-ams3c</RequestId><HostId>34e484cb-ams3c-ams3-zg03</HostId></Error>
```

After running a couple of times and hitting this error (possibly with different packages) the build stopped failing at this download stage and failed with a bad tar archive:

```
Error: TAR_BAD_ARCHIVE: Unrecognized archive format
    at DI.warn (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:882)
    at DI.warn (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:57168)
    at DI.<anonymous> (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:34247)
    at DI.emit (events.js:412:35)
    at DI.[emit] (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:36803)
    at DI.[maybeEnd] (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:38484)
    at DI.[consumeChunk] (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:38552)
    at DI.write (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:37972)
    at DI.end (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:39213)
    at HM.pipe (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:24711)
Emitted 'error' event on DI instance at:
    at DI.warn (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:855)
    at DI.warn (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:57168)
    [... lines matching original stack trace ...]
    at HM.pipe (file:///nix/store/sl649vdcccjpsgbvzfdai2scrii79wgf-node_spago-0.92.4/lib/node_modules/spago/bin/bundle.js:146:24711) {
  recoverable: false,
  file: '/home/nicholaw/.cache/spago-nodejs/packages/variant/8.0.0.tar.gz',
  cwd: '[...]/7368613235362d424c79313055734f78547572614a3547717a472f394b476753617872427a70593145313577446259342f593d',
  code: 'TAR_BAD_ARCHIVE',
  tarCode: 'TAR_BAD_ARCHIVE'
}
```

I presume either a failing download led to a bad archive being saved, or the download of a (perfectly fine) tarball was aborted/truncated on exit because of the failure in a different package. Either way the user cache becomes unusable

**Comment by f-f** [comment]
> the download of a (perfectly fine) tarball was aborted/truncated on exit because of the failure in a different package

Indeed this is what's happening: the download of all packages happens [in parallel](https://github.com/purescript/spago/blob/3933a81ed9b9606968f8c3285da27bd74d208c29/spaghetto/src/Spago/Command/Fetch.purs#L82), and the error you mention just [exits the program](https://github.com/purescript/spago/blob/3933a81ed9b9606968f8c3285da27bd74d208c29/spaghetto/src/Spago/Command/Fetch.purs#L114), likely interrupting [writing the tar to file](https://github.com/purescript/spago/blob/3933a81ed9b9606968f8c3285da27bd74d208c29/spaghetto/src/Spago/Command/Fetch.purs#L127) halfway through.

I suspect the best course of action here is to avoid calling `die` at all in this code block, and just return an array of `Either`. Exceptions might still be thrown in the underlying FFI code though, so probably worth to wrap them in a try/catch too

**Comment by f-f** [comment]
Eh, it's probably just easier to rearrange the logic so that we check (1) if the tar is there and (2) if we can extract it, and if not we just fetch a new one.

**Comment by nwolverson** [comment]
The latter approach of just checking when you try to use it is does seem like it would be more robust to any other issues

**Comment by f-f** [comment]
Fixed in 4f48741d1f74932762901e98489ef9f90f980f48
